src/items.rs

```diff
@@ -1262,6 +1262,55 @@ fn format_unit_struct(
     Some(format!("{}{};", header_str, generics_str))
 }
 
+fn set_brace_pos(
+    context: &RewriteContext<'_>,
+    struct_parts: &StructParts<'_>,
+    fields: &[ast::FieldDef],
+    body_lo: BytePos,
+) -> BracePos {
+    if fields.is_empty() {
+        let span = struct_parts.span;
+        let generics_lo = if let Some(generics) = struct_parts.generics {
+            generics.span.lo()
+        } else {
+            body_lo
+        };
+        let snippet = context.snippet(mk_sp(generics_lo, span.hi()));
+        let body_contains_comments = contains_comment(snippet);
+        let has_single_line_block_comment = if body_contains_comments {
+            let comment_start = match snippet.find('/') {
+                Some(i) => i,
+                None => 0,
+            };
+            let comment_snippet = &snippet[comment_start..];
+            let comment = comment_style(comment_snippet, false);
+            let is_block_comment = comment.is_block_comment();
+            let is_single_line_comment = if is_block_comment {
+                let comment_end = match find_comment_end(comment_snippet) {
+                    Some(i) => i,
+                    None => comment_start,
+                };
+                is_single_line(&comment_snippet[..=comment_end - 1])
+            } else {
+                false
+            };
+            is_single_line_comment
+        } else {
+            false
+        };
+
+        if context.config.version() == Version::Two {
+            return match body_contains_comments {
+                true if has_single_line_block_comment => BracePos::ForceSameLine,
+                true => BracePos::Auto,
+                false => BracePos::ForceSameLine,
+            };
+        }
+        return BracePos::ForceSameLine;
+    }
+    BracePos::Auto
+}
```

Comment:
I took a look, and I think we could simplify the logic here as follows:

**Note**: I renamed `set_brace_pos` to `set_struct_brace_pos` to be more explicit.
```rust
fn set_struct_brace_pos(
    context: &RewriteContext<'_>,
    fields: &[ast::FieldDef],
    span: Span,
) -> BracePos {
fn set_struct_brace_pos(
    context: &RewriteContext<'_>,
    fields: &[ast::FieldDef],
    span: Span,
) -> BracePos {
    if !fields.is_empty() {
        return BracePos::Auto;
    }

    // At this point we know there aren't any struct fields
    // This snippet contains the unformatted, trimmed struct body
    let snippet = context.snippet(span).trim();

    if snippet.is_empty() || context.config.version() == Version::One {
        return BracePos::ForceSameLine;
    }

    // The body isn't empty so there must be comments.
    // we also know that we're implicitly using Version Two formatting.
    let is_single_line = is_single_line(snippet);
    let is_block_comment = comment_style(snippet, false).is_block_comment();

    if is_single_line && is_block_comment {
        BracePos::ForceSameLine
    } else {
        BracePos::Auto
    }
}
```

At the call site we can hoist [`inner_span` from line 1329](https://github.com/rust-lang/rustfmt/blob/ef91154250977b3b5d05448dafbca524a1168b47/src/items.rs#L1329) so we can use that as the `span` argument to `set_struct_brace_pos`. Let me know if all that makes sense.